libp2p/src/builder.rs

```diff
@@ -0,0 +1,399 @@
+pub enum Tokio {}
+
+pub trait AuthenticatedMultiplexedTransport:
+    Transport<
+        Error = Self::E,
+        Dial = Self::D,
+        ListenerUpgrade = Self::U,
+        Output = (libp2p_identity::PeerId, StreamMuxerBox),
+    > + Send
+    + Unpin
+    + 'static
+{
+    type E: Send + Sync + 'static;
+    type D: Send;
+    type U: Send;
+}
+
+impl<T> AuthenticatedMultiplexedTransport for T
+where
+    T: Transport<Output = (libp2p_identity::PeerId, StreamMuxerBox)> + Send + Unpin + 'static,
+    <T as Transport>::Error: Send + Sync + 'static,
+    <T as Transport>::Dial: Send,
+    <T as Transport>::ListenerUpgrade: Send,
+{
+    type E = T::Error;
+    type D = T::Dial;
+    type U = T::ListenerUpgrade;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp"))]
+    fn tcp() {
+        let _: libp2p_swarm::Swarm<libp2p_swarm::dummy::Behaviour> = SwarmBuilder::new()
+            .with_new_identity()
+            .with_tokio()
+            .with_tcp()
+            .without_relay()
+            .no_more_other_transports()
+            .without_dns()
+            .with_behaviour(|_| libp2p_swarm::dummy::Behaviour)
+            .build();
+    }
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp", feature = "relay"))]
+    fn tcp_relay() {
+        #[derive(NetworkBehaviour)]
+        #[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+        struct Behaviour {
+            dummy: libp2p_swarm::dummy::Behaviour,
+            relay: libp2p_relay::client::Behaviour,
+        }
+
+        let _: libp2p_swarm::Swarm<Behaviour> = SwarmBuilder::new()
+            .with_new_identity()
+            .with_tokio()
+            .with_tcp()
+            .with_relay()
+            .no_more_other_transports()
+            .without_dns()
+            .with_behaviour(|_, relay| Behaviour {
+                dummy: libp2p_swarm::dummy::Behaviour,
+                relay,
+            })
+            .build();
+    }
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp", feature = "dns"))]
+    fn tcp_dns() {
+        let _: libp2p_swarm::Swarm<libp2p_swarm::dummy::Behaviour> = SwarmBuilder::new()
+            .with_new_identity()
+            .with_tokio()
+            .with_tcp()
+            .without_relay()
+            .no_more_other_transports()
+            .with_dns()
+            .with_behaviour(|_| libp2p_swarm::dummy::Behaviour)
+            .build();
+    }
+
+    /// Showcases how to provide custom transports unknown to the libp2p crate, e.g. QUIC or WebRTC.
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp"))]
+    fn tcp_other_transport_other_transport() {
+        let _: libp2p_swarm::Swarm<libp2p_swarm::dummy::Behaviour> = SwarmBuilder::new()
+            .with_new_identity()
+            .with_tokio()
+            .with_tcp()
+            .without_relay()
+            .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
```

Comment:
QUIC and WebRTC both will need access to the keypair so this should be a closure I'd say.

Reply:
:+1: done in https://github.com/libp2p/rust-libp2p/pull/4120/commits/343ba2d2729c4349544c4b027f8a2d6cc29e398b.